apis/v1alpha1/validation/validation_test.go

```diff
@@ -160,3 +160,275 @@ func TestValidateGateway(t *testing.T) {
+										Value: stringPtr("/"),
+									},
+								},
+							},
+							Filters: []gatewayv1a1.HTTPRouteFilter{
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestMirror,
+									RequestMirror: &gatewayv1a1.HTTPRequestMirrorFilter{
+										ServiceName: &testService,
+										Port:        portNumberPtr(8080),
+									},
+								},
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestHeaderModifier,
+									RequestHeaderModifier: &gatewayv1a1.HTTPRequestHeaderFilter{
+										Set: map[string]string{"special-header": "foo"},
+									},
+								},
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestMirror,
+									RequestMirror: &gatewayv1a1.HTTPRequestMirrorFilter{
+										ServiceName: &testService,
+										Port:        portNumberPtr(8080),
+									},
+								},
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestHeaderModifier,
+									RequestHeaderModifier: &gatewayv1a1.HTTPRequestHeaderFilter{
+										Add: map[string]string{"my-header": "bar"},
+									},
+								},
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestMirror,
+									RequestMirror: &gatewayv1a1.HTTPRequestMirrorFilter{
+										ServiceName: &specialService,
+										Port:        portNumberPtr(8080),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			errCount: 2,
+		},
+		{
+			name: "valid httpRoute with duplicate ExtensionRef filters",
+			hRoute: gatewayv1a1.HTTPRoute{
+				Spec: gatewayv1a1.HTTPRouteSpec{
+					Rules: []gatewayv1a1.HTTPRouteRule{
+						{
+							Matches: []gatewayv1a1.HTTPRouteMatch{
+								{
+									Path: &gatewayv1a1.HTTPPathMatch{
+										Type:  pathMatchTypePtr("Prefix"),
+										Value: stringPtr("/"),
+									},
+								},
+							},
+							Filters: []gatewayv1a1.HTTPRouteFilter{
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestHeaderModifier,
+									RequestHeaderModifier: &gatewayv1a1.HTTPRequestHeaderFilter{
+										Set: map[string]string{"special-header": "foo"},
+									},
+								},
+								{
+									Type: gatewayv1a1.HTTPRouteFilterRequestMirror,
+									RequestMirror: &gatewayv1a1.HTTPRequestMirrorFilter{
+										ServiceName: &testService,
+										Port:        portNumberPtr(8080),
+									},
+								},
+								{
+									Type: "ExtensionRef",
+								},
+								{
+									Type: "ExtensionRef",
+								},
+								{
+									Type: "ExtensionRef",
+								},
+							},
+						},
+					},
+				},
+			},
+			errCount: 0,
+		},
+	}
+	for _, tt := range tests {
+		// copy variable to avoid scope problems with ranges
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			errs := ValidateHTTPRoute(&tt.hRoute)
+			if len(errs) != tt.errCount {
+				t.Errorf("ValidateHTTPRoute() got %v errors, want %v errors", len(errs), tt.errCount)
+			}
+		})
+	}
+}
+
+func stringPtr(s string) *string {
+	return &s
+}
+
+func int32Ptr(i int) *int32 {
+	result := int32(i)
+	return &result
+}
```

Comment:
Not sure if we want to add a dependency here, but this is largely similar to https://github.com/kubernetes/utils/tree/master/pointer.